RFCs/0031-Multi-Test-Run-Finalization.md

```diff
@@ -0,0 +1,137 @@
+    /// </summary>
+    public interface IMultiTestRunFinalizationEventsHandler : ITestMessageEventHandler
+    {
+        /// <summary>
+        /// Dispatch MultiTestRunFinalizationComplete event to listeners.
+        /// </summary>
+        /// <param name="finalizationCompleteEventArgs">Finalization Complete event args.</param>
+        /// <param name="attachments">Last set of processed attachment sets.</param>
+        void HandleMultiTestRunFinalizationComplete(MultiTestRunFinalizationCompleteEventArgs finalizationCompleteEventArgs, IEnumerable<AttachmentSet> lastChunk);
+
+        /// <summary>
+        /// Dispatch FinalisedAttachments event to listeners.
+        /// </summary>
+        /// <param name="attachments">Finalised attachment sets.</param>
+        void HandleFinalisedAttachments(IEnumerable<AttachmentSet> attachments);
+
+        /// <summary>
+        /// Dispatch MultiTestRunFinalizationProgress event to listeners.
+        /// </summary>
+        /// <param name="finalizationProgressEventArgs">Finalization Progress event args.</param>
+        void HandleMultiTestRunFinalizationProgress(MultiTestRunFinalizationProgressEventArgs finalizationProgressEventArgs);
+    }
+}
+```
+Interface provides callbacks from Multi Test Run Finalization process. For every finalization process `HandleMultiTestRunFinalizationComplete` will be called once and will provide last chunk or all data attachments. During finalization process `HandleFinalisedAttachments` can be invoked several times providing data attachments that are already processed. Method `HandleMultiTestRunFinalizationProgress` will be invoked every time when `progressReporter` is used by `IDataCollectorAttachments` implementation and will provide information about current collector: progress, uris and index of collector. Additionally event will contain also number of collectors.
+
+
+4. Use above logic to combine/merge data attachments for parallel test executions and VS scenarios (e.g. `Run All Tests`, `Analyze Code Coverage for All Tests`). In case of `Analyze Code Coverage for All Tests` VS will use `vstest.console` in Design Mode and merge all code coverage reports. VS will show full code coverage report for all test projects.
+
+5. When [dotnet test](https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet-test) command is used to execute unit tests in a given solution, new console app `Orchestrator` will be executed.
+  * `Orchestrator` for every project inside solution will start `dotnet test` command, using `Process` with output redirected. Output for every project will be printed to stdout with some mutex to not mix output from children. 
+  * Additionally new test platform extension will be introduced and will propagate data about:
+    - Attachments
+    - Test results statistics
+  back to `Orchestrator`.
+  * In parallel with test executions `Orchestrator` will start `vstest.console` in Design Mode.
```

Comment:
`in Design mode` -> `in a variation of design mode`

Reply:
I'm assuming we wouldn't be sending back test case and test result objects in this mode. Do we know what this interface will look like?

Reply:
I would create new interface for this. It could be similar to:

interface IVsTestConsoleListener
{
     public Task ListenForTestRunEventsAsync(ITestRunEventsHandler testRunEventsHandler);
}

another possibility that maybe we can execute for dotnet everything in design mode. In both cases its not changing existing Multi Test Run Finalization design. I would not include it in this document. WDYT?

Reply:
Yes, this document is primarily focused on flushing out the IDE scenario. The CLI workflows I'd expect to be more detailed in another doc based on the overview you've detailed here, when we get to that work. As long as we've thought about how these specific data collector attachment processing interfaces would apply to CLI workflows I think this document would do justice and I think we've achieved that. Perhaps that's something you'd want to add to the summary of this doc.

---
In reply to: [447004623](https://github.com/microsoft/vstest-docs/pull/235#discussion_r447004623) [](ancestors = 447004623)

Reply:
WDYT about moving 'dotnet test' specific things to separated document and keep it on opened PR? This document could focus only on finalization process, not so much on using it?

Reply:
Sounds good to me

Reply:
Moved to PR: https://github.com/microsoft/vstest-docs/pull/236